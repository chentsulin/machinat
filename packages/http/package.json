{
  "name": "@machinat/http",
  "version": "0.4.0-beta.1",
  "author": "Patrick Liu <zeldalink0515@gmail.com>  (https://github.com/lrills)",
  "repository": {
    "type": "git",
    "url": "https://github.com/machinat/machinat.git",
    "directory": "packages/http"
  },
  "homepage": "https://github.com/machinat/machinat#readme",
  "bugs": "https://github.com/machinat/machinat/issues",
  "main": "lib/index.js",
  "exports": {
    ".": "./lib/index.js",
    "./types": "./lib/types.js",
    "./webhook": "./lib/webhook/index.js",
    "./webhook/types": "./lib/webhook/types.js"
  },
  "files": [
    "lib/*",
    "README.md",
    "polyfill-exports.js"
  ],
  "scripts": {
    "postinstall": "./polyfill-exports.js"
  },
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "invariant": "^2.2.4",
    "raw-body": "^2.4.1",
    "thenifiedly": "^0.2.1"
  },
  "peerDependencies": {
    "@machinat/core": "^0.4.1-alpha"
  }
}
