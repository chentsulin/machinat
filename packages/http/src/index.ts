export { default } from './module';
export { HttpConnector } from './connector';
export * from './types';
