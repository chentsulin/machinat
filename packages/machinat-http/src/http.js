// @flow
import url from 'url';
import type { IncomingMessage, ServerResponse } from 'http';
import { HTTPReceivable } from './types';

export const connect = (bot: HTTPReceivable) => (
  req: IncomingMessage,
  res: ServerResponse
) => {
  bot.adaptor.handleRequest(req, res);
};

export const connectByRoutes = (mapping: { [string]: HTTPReceivable }) => {
  const routes = Object.keys(mapping);

  return (req: IncomingMessage, res: ServerResponse) => {
    const parsed = url.parse(req.url);
    const pathname = parsed.pathname || '';

    for (const route of routes) {
      if (pathname.slice(-route.length) === route) {
        mapping[route].adaptor.handleRequest(req, res);
        return;
      }
    }

    res.statusCode = 404; // eslint-disable-line no-param-reassign
    res.end();
  };
};
