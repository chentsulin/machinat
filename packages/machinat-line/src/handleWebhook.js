// @flow
import crypto from 'crypto';
import type { IncomingMessage, ServerResponse } from 'http';
import type {
  WebhookHandler,
  WebhookEventReport,
} from 'machinat-webhook-receiver/types';

import createEvent from './event';
import LineThread from './thread';

import type {
  LineBotOptions,
  LineWebhookRequestBody,
  LineEvent,
} from './types';

const endRes = (res, code, body) => {
  res.statusCode = code; // eslint-disable-line no-param-reassign
  res.end(body);
};

const handleWebhook = (
  options: LineBotOptions
): WebhookHandler<LineThread, LineEvent> => (
  req: IncomingMessage,
  res: ServerResponse,
  rawBody: void | string
) => {
  if (req.method !== 'POST') {
    return endRes(res, 405);
  }

  if (rawBody === undefined) {
    return endRes(res, 400);
  }

  const { shouldValidateRequest, channelSecret } = options;

  // NOTE: channelSecret is validated at bot ðŸ‘‡
  if (shouldValidateRequest /* :: && channelSecret */) {
    const signature = crypto
      .createHmac('SHA256', channelSecret)
      .update(rawBody)
      .digest('base64');

    if (req.headers['x-line-signature'] !== signature) {
      return endRes(res, 401);
    }
  }

  try {
    const { events: payloads } = (JSON.parse(rawBody): LineWebhookRequestBody);

    const events: WebhookEventReport<LineThread, LineEvent>[] = new Array(
      payloads.length
    );

    for (let i = 0; i < events.length; i += 1) {
      const payload = payloads[i];

      const event = createEvent(payload);
      const thread = new LineThread(payload.source);

      events[i] = { event, thread, shouldRespond: false };
    }

    return events;
  } catch (e) {
    return endRes(res, 400);
  }
};

export default handleWebhook;
