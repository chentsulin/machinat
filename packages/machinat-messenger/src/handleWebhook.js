// @flow
import url from 'url';
import crypto from 'crypto';

import type { IncomingMessage, ServerResponse } from 'http';
import type {
  WebhookHandler,
  WebhookEventReport,
} from 'machinat-webhook-receiver/types';
import type { MessengerBotOptions, MessengerEvent } from './types';

import createEvent from './event';
import MessengerChannel from './channel';

const endRes = (res, code, body) => {
  res.statusCode = code; // eslint-disable-line no-param-reassign
  res.end(body);
};

const handleWebhook = (
  options: MessengerBotOptions
): WebhookHandler<MessengerChannel, MessengerEvent> => (
  req: IncomingMessage,
  res: ServerResponse,
  rawBody?: string
) => {
  const {
    shouldVerifyWebhook,
    verifyToken,
    shouldValidateRequest,
    appSecret,
  } = options;

  if (req.method === 'GET') {
    const query: { [string]: string } = (url.parse(req.url, true).query: any);

    if (!shouldVerifyWebhook) {
      endRes(res, 403);
    } else if (
      query['hub.mode'] === 'subscribe' &&
      query['hub.verify_token'] === verifyToken
    ) {
      endRes(res, 200, query['hub.challenge']);
    } else {
      endRes(res, 400);
    }

    return null;
  }

  if (req.method !== 'POST') {
    endRes(res, 405);
    return null;
  }

  if (!rawBody) {
    endRes(res, 400);
    return null;
  }

  if (shouldValidateRequest && appSecret !== undefined) {
    const hmac = crypto.createHmac('sha1', appSecret);

    hmac.update(rawBody, 'utf8');
    const computedSig = `sha1=${hmac.digest('hex')}`;

    if (req.headers['x-hub-signature'] !== computedSig) {
      endRes(res, 401);
      return null;
    }
  }

  let body;
  try {
    body = JSON.parse(rawBody);
  } catch (e) {
    endRes(res, 400);
    return null;
  }

  if (body.object !== 'page') {
    endRes(res, 404);
    return null;
  }

  const reports: WebhookEventReport<MessengerChannel, MessengerEvent>[] = [];

  for (const { messaging, stanby } of body.entry) {
    const isStandby = stanby !== undefined;
    const rawEvents = isStandby ? stanby : messaging;

    for (const raw of rawEvents) {
      const event = createEvent(isStandby, raw);
      const { type, payload } = event;

      const source =
        type === 'optin' && payload.sender === undefined
          ? { user_ref: payload.optin.user_ref }
          : payload.sender;
      const channel = new MessengerChannel(source);

      const shouldRespond =
        type === 'checkout_update' || type === 'pre_checkout';

      reports.push({ event, channel, shouldRespond });
    }
  }

  return reports;
};

export default handleWebhook;
