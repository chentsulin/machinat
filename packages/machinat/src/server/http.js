import http from 'http';

const createServer = connector => {
  const handleWebhook = connector.handleWebhook();
  const empty = () => {};

  return http.createServer((req, res) => {
    handleWebhook(req, res).catch(empty);
  });
};

export default createServer;
