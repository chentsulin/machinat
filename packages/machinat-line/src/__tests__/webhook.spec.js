import { IncomingMessage, ServerResponse } from 'http';
import moxy from 'moxy';

import handleWebhook from '../webhook';
import LineChannel from '../channel';

describe('handleWebhook(options)(req, res, body)', () => {
  it.each(['GET', 'PUT', 'PATCH', 'DELETE', 'HEAD', 'UPDATE', 'UPGRADE'])(
    'responds 405 if req.method is %s',
    async method => {
      const req = moxy(new IncomingMessage());
      req.method = method;
      const res = moxy(new ServerResponse({ method }));

      await expect(
        handleWebhook({
          channelId: '_LINE_CHANNEL_ID_',
          shouldValidateRequest: false,
        })(req, res, 'body')
      ).resolves.toBe(null);

      expect(res.statusCode).toBe(405);
      expect(res.finished).toBe(true);
    }
  );

  it('responds 400 if body is empty', async () => {
    const req = moxy(new IncomingMessage());
    req.method = 'POST';
    const res = moxy(new ServerResponse({ method: 'POST' }));

    await expect(
      handleWebhook({
        channelId: '_LINE_CHANNEL_ID_',
        shouldValidateRequest: false,
      })(req, res)
    ).resolves.toBe(null);

    expect(res.statusCode).toBe(400);
    expect(res.finished).toBe(true);
  });

  it('responds 400 if body is not not valid json format', async () => {
    const req = moxy(new IncomingMessage());
    req.method = 'POST';
    const res = moxy(new ServerResponse({ method: 'POST' }));

    await expect(
      handleWebhook({
        channelId: '_LINE_CHANNEL_ID_',
        shouldValidateRequest: false,
      })(req, res, '_invalid_body_')
    ).resolves.toBe(null);

    expect(res.statusCode).toBe(400);
    expect(res.finished).toBe(true);
  });

  it('responds 400 if body is in invalid format', async () => {
    const req = moxy(new IncomingMessage());
    req.method = 'POST';
    const res = moxy(new ServerResponse({ method: 'POST' }));

    await expect(
      handleWebhook({
        channelId: '_LINE_CHANNEL_ID_',
        shouldValidateRequest: false,
      })(req, res, JSON.stringify({ there: 'is no events hahaha' }))
    ).resolves.toBe(null);

    expect(res.statusCode).toBe(400);
    expect(res.finished).toBe(true);
  });

  it('returns events created', async () => {
    const req = moxy(new IncomingMessage());
    req.method = 'POST';
    const res = moxy(new ServerResponse({ method: 'POST' }));

    const body = {
      destination: 'xxxxxxxxxx',
      events: [
        {
          replyToken: '0f3779fba3b349968c5d07db31eab56f',
          type: 'message',
          timestamp: 1462629479859,
          source: {
            type: 'user',
            userId: 'U4af4980629',
          },
          message: {
            id: '325708',
            type: 'text',
            text: 'Hello, world',
          },
        },
        {
          replyToken: '8cf9239d56244f4197887e939187e19e',
          type: 'follow',
          timestamp: 1462629479859,
          source: {
            type: 'user',
            userId: 'U4af4980629',
          },
        },
      ],
    };

    const eventResults = await handleWebhook({
      channelId: '_LINE_CHANNEL_ID_',
      shouldValidateRequest: false,
    })(req, res, JSON.stringify(body));

    expect(res.statusCode).toBe(200);
    expect(res.finished).toBe(true);

    expect(eventResults.length).toBe(2);

    expect(eventResults).toMatchSnapshot();

    eventResults.forEach(({ event, channel, response }, i) => {
      expect(response).toBe(undefined);

      expect(channel).toBeInstanceOf(LineChannel);
      expect(channel.subtype).toBe('user');
      expect(channel.source).toEqual({
        type: 'user',
        userId: 'U4af4980629',
      });
      expect(channel.uid).toBe('line:_LINE_CHANNEL_ID_:user:U4af4980629');

      expect(event.platform).toBe('line');
      expect(event.type).toBe(!i ? 'message' : 'follow');
      expect(event.subtype).toBe(!i ? 'text' : undefined);
      expect(event.payload).toEqual(body.events[i]);
    });
  });

  it('pass options.channelId to channel', async () => {
    const req = moxy(new IncomingMessage());
    req.method = 'POST';
    const res = moxy(new ServerResponse({ method: 'POST' }));

    const [{ channel }] = await handleWebhook({
      shouldValidateRequest: false,
      channelId: '_my_foo_line_channel_',
    })(
      req,
      res,
      JSON.stringify({
        destination: 'xxxxxxxxxx',
        events: [
          {
            replyToken: '0f3779fba3b349968c5d07db31eab56f',
            type: 'message',
            timestamp: 1462629479859,
            source: { type: 'user', userId: 'U4af4980629' },
            message: { id: '325708', type: 'text', text: 'Hello, world' },
          },
        ],
      })
    );

    expect(channel).toBeInstanceOf(LineChannel);
    expect(channel.subtype).toBe('user');
    expect(channel.uid).toBe('line:_my_foo_line_channel_:user:U4af4980629');
  });

  it('returns events if request validation passed', async () => {
    const req = moxy(new IncomingMessage());
    req.method = 'POST';
    const res = moxy(new ServerResponse({ method: 'POST' }));

    const body =
      '{"destination":"xxx","events":[{"replyToken":"xxx","type":"message","timestamp":1462629479859,"source":{"type":"user","userId":"xxx"},"message":{"id":"325708","type":"text","text":"Hello, world"}}]}';
    const hmac = 'tJIFpuDMyZ4a9XzwkNUBK2B/7NH5gxYJDR/57RCf+6k=';
    req.mock.getter('headers').fakeReturnValue({ 'x-line-signature': hmac });

    const [{ event, channel, response }] = await handleWebhook({
      channelId: '_LINE_CHANNEL_ID_',
      shouldValidateRequest: true,
      channelSecret: '__LINE_CHANNEL_SECRET__',
    })(req, res, body);

    expect(response).toBe(undefined);

    expect(res.statusCode).toBe(200);
    expect(res.finished).toBe(true);

    expect(channel.source).toEqual({ type: 'user', userId: 'xxx' });

    expect(event.platform).toBe('line');
    expect(event.type).toBe('message');
    expect(event.subtype).toBe('text');
    expect(event.payload).toEqual({
      replyToken: 'xxx',
      type: 'message',
      timestamp: 1462629479859,
      source: { type: 'user', userId: 'xxx' },
      message: { id: '325708', type: 'text', text: 'Hello, world' },
    });
  });

  it('responds 401 if request validation failed', async () => {
    const req = moxy(new IncomingMessage());
    req.method = 'POST';
    const res = moxy(new ServerResponse({ method: 'POST' }));

    const body =
      '{"destination":"xxx","events":[{"replyToken":"xxx","type":"message","timestamp":1462629479859,"source":{"type":"user","userId":"xxx"},"message":{"id":"325708","type":"text","text":"Hello, world"}}]}';
    const hmac = '__INVALID_SIGANATURE__';
    req.mock.getter('headers').fakeReturnValue({ 'x-line-signature': hmac });

    await expect(
      handleWebhook({
        channelId: '_LINE_CHANNEL_ID_',
        shouldValidateRequest: true,
        channelSecret: '__LINE_CHANNEL_SECRET__',
      })(req, res, body)
    ).resolves.toBe(null);

    expect(res.statusCode).toBe(401);
    expect(res.finished).toBe(true);
  });
});
