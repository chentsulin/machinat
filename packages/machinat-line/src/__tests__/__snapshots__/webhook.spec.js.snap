// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleWebhook(options)(req, res, body) returns events created 1`] = `
Array [
  Object {
    "channel": LineChannel {
      "channelId": "_LINE_CHANNEL_ID_",
      "platform": "line",
      "source": Object {
        "type": "user",
        "userId": "U4af4980629",
      },
      "sourceId": "U4af4980629",
      "type": "user",
    },
    "event": Object {
      "payload": Object {
        "message": Object {
          "id": "325708",
          "text": "Hello, world",
          "type": "text",
        },
        "replyToken": "0f3779fba3b349968c5d07db31eab56f",
        "source": Object {
          "type": "user",
          "userId": "U4af4980629",
        },
        "timestamp": 1462629479859,
        "type": "message",
      },
      "subtype": "text",
      "type": "message",
    },
    "response": undefined,
    "user": LineUser {
      "platform": "line",
      "userId": "U4af4980629",
    },
  },
  Object {
    "channel": LineChannel {
      "channelId": "_LINE_CHANNEL_ID_",
      "platform": "line",
      "source": Object {
        "type": "user",
        "userId": "U4af4980629",
      },
      "sourceId": "U4af4980629",
      "type": "user",
    },
    "event": Object {
      "payload": Object {
        "replyToken": "8cf9239d56244f4197887e939187e19e",
        "source": Object {
          "type": "user",
          "userId": "U4af4980629",
        },
        "timestamp": 1462629479859,
        "type": "follow",
      },
      "subtype": undefined,
      "type": "follow",
    },
    "response": undefined,
    "user": LineUser {
      "platform": "line",
      "userId": "U4af4980629",
    },
  },
]
`;
